arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -804,6 +804,10 @@ ErrorCode DatabaseFeature::dropDatabase(std::string_view name) {
 
     TRI_vocbase_t* vocbase = it->second;
 
+    // Signal replicated logs to stop here, while they still have access to the
+    // database.
+    vocbase->shutdownReplicatedLogs();
+
```

Comment:
Doing it this early will actually introduce a race: New replicated logs could still be added, which are then not shutdown at all. The call to shutdown logs must only happen after the vocbase was marked as dropped, which prevents new logs from being created.

Reply:
To my current understanding, no replicated states (nor logs) may be created after `VocbaseLogManager::resignAll()` is called. `shutdownReplicatedLogs()` just calls into `_logManager->resignAll()`.
The `resignAll()` starts with the following lines:
```
auto guard = _guardedData.getLockedGuard();
guard->resignAllWasCalled = true;
``` 
The `resignAllWasCalled` atomic is used to prevent new instances from being created.
The one point of creation (that I am aware of) for replicated logs and states is `VocBaseLogManager::GuardedData::buildReplicatedStateWithMethods()`. This does the following check in the beginning:
```
if (resignAllWasCalled) {
  return {TRI_ERROR_SHUTTING_DOWN};
}
```
I think this is enough to prevent the race you're mentioning.

Reply:
Ah, good! Thanks for the explanation. The error message is probably wrong when a database is being deleted, rather than the process being shut down, but that doesn't have to be fixed here.